### DISCUSSION TITLE
[Question] How to use Tekton in KubeVela | 请问我该如何在 KubeVela 中使用 Tekton

### DISCUSSION THREAD

**Discussion by cnbailian** [kubevela/kubevela#1755] [Q&A]
**问题描述**
我想在 KubeVela 中使用 Tekton 来完成 CI，最初的想法是 Tekton 的各个 CRD 会作为 Component 在 Application 中被组合，最后通过 Trait 部署另一个 Appplication 来达成 CD 的需求。但这个想法有很多不完善的地方，比如一个应用会有两个 Application，一个是 CI，一个是真正的应用。而且描述文件会显得有些混乱。

在钉钉群里请教之后，得知新特性 Workflow 可能会匹配这个场景，所以我在发现有相关 PR 后尝试了一下，发现也不能解决我的需求。主要的困难是不能在 Workflow 中输出没有状态的资源。Tekton 的很多组件都是没有状态的，只是等待 TaskRun 等资源应用时读取，所以我没有办法在 Application.Workflow 中使用 Tekton。
在 PR 中提出了我的困惑后，hongchaodeng 大佬告诉我这里应该使用 Component 和 Trait。但我在 Component 中创建的 Tekton Task 资源，只是保存在了 AppRevision 中，无法让 Workflow 中创建的 TaskRun 读取到。

请问我的这个使用场景是不是不符合 workflow 的适用场景。目前在 KubeVela 中是否有将 CI 组件组合的例子呢？

**相关资源**
component + trait example：https://github.com/cnbailian/kubevela-examples-go/tree/master/docs/examples/cd
workflow example：https://github.com/cnbailian/kubevela-examples-go/tree/master/docs/examples/workflow-ci

**Comment by hongchaodeng** [comment]
@cnbailian 你好。感谢你的提问，这次的描述非常具体，让我也有更仔细的思考。

个人认为，如果是 CICD 的场景，是需要用 Workflow 来描述的。首先我先指出一点，那就是 Tekton TaskRun 等组件并不是你所说的无状态的。它也有 [conditions 和 steps 等状态](https://tekton.dev/docs/pipelines/taskruns/) 来汇报运行进展。这里面问题在于 Tekton native resource 像 TaskRun 等没法适配 Workflow 里对 status.conditions 的要求。为此，我们需要额外再加一个 adaptor CRD，来将 TaskRun 等信息包住，并把状态 translate 成 Workflow 能够理解的信息。

**Comment by cnbailian** [comment]
@hongchaodeng 感谢回复，同时也感谢您指出我的错误。我上面的表述不清晰，我想说的不是 Tekton 的 TaksRun 组件无状态，而是说 TaskRun 所引用的 Task 和 PipelineResource 等资源是无状态的，它们只有 Spec，没有 Status 结构。在这种情况下我没有办法让这些资源在 Workflow 中创建。如果是在 trait 或 component 中，也没有办法让在 Workflow 中创建的 TaskRun 资源引用到，因为它们此时还没有创建。
这是我目前所遇到的困境。

**Comment by hongchaodeng** [comment]
@cnbailian 

> TaskRun 所引用的 Task 和 PipelineResource 等资源是无状态的，它们只有 Spec，没有 Status 结构。

如果是这样的话，你只需要把他们 model 成一个 component 或者 policy。然后 workflow 里面跑两步：1. 第一步是把这些 CI steps 跑起来；2. 第二步才是部署真正的 Application 里的资源。这个需要定制化 workflow 里的 apply 操作。

**Comment by cnbailian** [comment]
@hongchaodeng 也就是说我现在的使用方式是有问题的，我要在 workflow 第一步输出一个 application（包含 CI 的全部资源） 用于执行 CI？而不是在这个 Application 资源本身的 component 中设置。然后第二步是输出真正的用于部署的 appdeployment。
这样做合适吗？输出 application 作为 CI 的执行抽象。

**Reply by hongchaodeng** [reply]
> 我要在 workflow 第一步输出一个 application（包含 CI 的全部资源） 用于执行 CI？

这个在 Vela 里实现，全部资源会放在 ConfigMap 里。

**Reply by hongchaodeng** [reply]
是的。Application 作为应用交付的唯一入口，将所有交付相关的操作都描述进来，以统一的方式交付。

**Reply by cnbailian** [reply]
> 这个在 Vela 里实现，全部资源会放在 ConfigMap 里。

Vela 里是指 Vela Controllers 中？这是后面的开发计划吗
我理想中的情况是在一个 Application 中描述交付的所有操作，并且不想在 workflow 或 component 中输出另一个 Application，这样会对我的用户造成混乱。我可以期待一下 workflow 的下一步开发进度吗

**Reply by hongchaodeng** [reply]
是的。
详情请看这个设计文档：https://github.com/oam-dev/kubevela/blob/master/design/vela-core/workflow_policy.md#technical-details
